In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
from sklearn.model_selection import train_test_split

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/-spotify-tracks-dataset/dataset.csv


In [2]:
data = pd.read_csv('/kaggle/input/-spotify-tracks-dataset/dataset.csv')
df = pd.DataFrame(data)
df

Unnamed: 0                track_id                 artists  \
0                0  5SuOikwiRyPMVoIQDJUgSV             Gen Hoshino   
1                1  4qPNDBW1i3p13qLCt0Ki3A            Ben Woodward   
2                2  1iJBSr7s7jYXzM8EGcbK5b  Ingrid Michaelson;ZAYN   
3                3  6lfxq3CG4xtTiEg7opyCyx            Kina Grannis   
4                4  5vjLSffimiIP26QG5WcN2K        Chord Overstreet   
...            ...                     ...                     ...   
113995      113995  2C3TZjDRiAzdyViavDJ217           Rainy Lullaby   
113996      113996  1hIz5L4IB9hN3WRYPOCGPw           Rainy Lullaby   
113997      113997  6x8ZfSoqDjuNa5SVP5QjvX           Cesária Evora   
113998      113998  2e6sXL2bYv4bSz6VTdnfLs        Michael W. Smith   
113999      113999  2hETkH7cOfqmz3LqZDHZf5           Cesária Evora   

                                               album_name  \
0                                                  Comedy   
1                                        Ghost (Acoustic)   
2                                          To Begin Again   
3       Crazy Rich Asians (Original Motion Picture Sou...   
4                                                 Hold On   
...                                                   ...   
113995  #mindfulness - Soft Rain for Mindful Meditatio...   
113996  #mindfulness - Soft Rain for Mindful Meditatio...   
113997                                            Best Of   
113998                                  Change Your World   
113999                                     Miss Perfumado   

                        track_name  popularity  duration_ms  explicit  \
0                           Comedy          73       230666     False   
1                 Ghost - Acoustic          55       149610     False   
2                   To Begin Again          57       210826     False   
3       Can't Help Falling In Love          71       201933     False   
4                          Hold On          82       198853     False   
...                            ...         ...          ...       ...   
113995         Sleep My Little Boy          21       384999     False   
113996            Water Into Light          22       385000     False   
113997              Miss Perfumado          22       271466     False   
113998                     Friends          41       283893     False   
113999                   Barbincor          22       241826     False   

        danceability  energy  ...  loudness  mode  speechiness  acousticness  \
0              0.676  0.4610  ...    -6.746     0       0.1430        0.0322   
1              0.420  0.1660  ...   -17.235     1       0.0763        0.9240   
2              0.438  0.3590  ...    -9.734     1       0.0557        0.2100   
3              0.266  0.0596  ...   -18.515     1       0.0363        0.9050   
4              0.618  0.4430  ...    -9.681     1       0.0526        0.4690   
...              ...     ...  ...       ...   ...          ...           ...   
113995         0.172  0.2350  ...   -16.393     1       0.0422        0.6400   
113996         0.174  0.1170  ...   -18.318     0       0.0401        0.9940   
113997         0.629  0.3290  ...   -10.895     0       0.0420        0.8670   
113998         0.587  0.5060  ...   -10.889     1       0.0297        0.3810   
113999         0.526  0.4870  ...   -10.204     0       0.0725        0.6810   

        instrumentalness  liveness  valence    tempo  time_signature  \
0               0.000001    0.3580   0.7150   87.917               4   
1               0.000006    0.1010   0.2670   77.489               4   
2               0.000000    0.1170   0.1200   76.332               4   
3               0.000071    0.1320   0.1430  181.740               3   
4               0.000000    0.0829   0.1670  119.949               4   
...                  ...       ...      ...      ...             ...   
113995          0.928000    0.0863   0.0339  125.995               5   
113996          0.97

**Exploratory Data Analysis**

In [3]:
# Clean the data (dropping NA values)
df.isna().sum()
df = df.dropna()

In [4]:
# Data description and distributions
df.describe()

Unnamed: 0     popularity   duration_ms   danceability  \
count  113999.000000  113999.000000  1.139990e+05  113999.000000   
mean    56999.421925      33.238827  2.280312e+05       0.566801   
std     32909.243463      22.304959  1.072961e+05       0.173543   
min         0.000000       0.000000  8.586000e+03       0.000000   
25%     28499.500000      17.000000  1.740660e+05       0.456000   
50%     56999.000000      35.000000  2.129060e+05       0.580000   
75%     85499.500000      50.000000  2.615060e+05       0.695000   
max    113999.000000     100.000000  5.237295e+06       0.985000   

              energy            key       loudness           mode  \
count  113999.000000  113999.000000  113999.000000  113999.000000   
mean        0.641383       5.309126      -8.258950       0.637558   
std         0.251530       3.559999       5.029357       0.480708   
min         0.000000       0.000000     -49.531000       0.000000   
25%         0.472000       2.000000     -10.013000       0.000000   
50%         0.685000       5.000000      -7.004000       1.000000   
75%         0.854000       8.000000      -5.003000       1.000000   
max         1.000000      11.000000       4.532000       1.000000   

         speechiness   acousticness  instrumentalness       liveness  \
count  113999.000000  113999.000000     113999.000000  113999.000000   
mean        0.084652       0.314907          0.156051       0.213554   
std         0.105733       0.332522          0.309556       0.190378   
min         0.000000       0.000000          0.000000       0.000000   
25%         0.035900       0.016900          0.000000       0.098000   
50%         0.048900       0.169000          0.000042       0.132000   
75%         0.084500       0.597500          0.049000       0.273000   
max         0.965000       0.996000          1.000000       1.000000   

             valence          tempo  time_signature  
count  113999.000000  113999.000000   113999.000000  
mean        0.474066     122.147695        3.904034  
std         0.259261      29.978290        0.432623  
min         0.000000       0.000000        0.000000  
25%         0.260000      99.218500        4.000000  
50%         0.464000     122.017000        4.000000  
75%         0.683000     140.071000        4.000000  
max         0.995000     243.372000        5.000000

In [5]:
# Non-quantitative columns
df.describe(include=['O'])

track_id      artists                  album_name  \
count                   113999       113999                      113999   
unique                   89740        31437                       46589   
top     6S3JlDAGk3uu3NtZbPnuhS  The Beatles  Alternative Christmas 2022   
freq                         9          279                         195   

             track_name track_genre  
count            113999      113999  
unique            73608         114  
top     Run Rudolph Run    acoustic  
freq                151        1000

From this, we can see that there are several duplicated tracks. We will remove them as they are not neccessary.

In [6]:
df.drop_duplicates(subset=['track_id'], inplace=True)
df

/tmp/ipykernel_21/2910389361.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.drop_duplicates(subset=['track_id'], inplace=True)


Unnamed: 0                track_id                 artists  \
0                0  5SuOikwiRyPMVoIQDJUgSV             Gen Hoshino   
1                1  4qPNDBW1i3p13qLCt0Ki3A            Ben Woodward   
2                2  1iJBSr7s7jYXzM8EGcbK5b  Ingrid Michaelson;ZAYN   
3                3  6lfxq3CG4xtTiEg7opyCyx            Kina Grannis   
4                4  5vjLSffimiIP26QG5WcN2K        Chord Overstreet   
...            ...                     ...                     ...   
113995      113995  2C3TZjDRiAzdyViavDJ217           Rainy Lullaby   
113996      113996  1hIz5L4IB9hN3WRYPOCGPw           Rainy Lullaby   
113997      113997  6x8ZfSoqDjuNa5SVP5QjvX           Cesária Evora   
113998      113998  2e6sXL2bYv4bSz6VTdnfLs        Michael W. Smith   
113999      113999  2hETkH7cOfqmz3LqZDHZf5           Cesária Evora   

                                               album_name  \
0                                                  Comedy   
1                                        Ghost (Acoustic)   
2                                          To Begin Again   
3       Crazy Rich Asians (Original Motion Picture Sou...   
4                                                 Hold On   
...                                                   ...   
113995  #mindfulness - Soft Rain for Mindful Meditatio...   
113996  #mindfulness - Soft Rain for Mindful Meditatio...   
113997                                            Best Of   
113998                                  Change Your World   
113999                                     Miss Perfumado   

                        track_name  popularity  duration_ms  explicit  \
0                           Comedy          73       230666     False   
1                 Ghost - Acoustic          55       149610     False   
2                   To Begin Again          57       210826     False   
3       Can't Help Falling In Love          71       201933     False   
4                          Hold On          82       198853     False   
...                            ...         ...          ...       ...   
113995         Sleep My Little Boy          21       384999     False   
113996            Water Into Light          22       385000     False   
113997              Miss Perfumado          22       271466     False   
113998                     Friends          41       283893     False   
113999                   Barbincor          22       241826     False   

        danceability  energy  ...  loudness  mode  speechiness  acousticness  \
0              0.676  0.4610  ...    -6.746     0       0.1430        0.0322   
1              0.420  0.1660  ...   -17.235     1       0.0763        0.9240   
2              0.438  0.3590  ...    -9.734     1       0.0557        0.2100   
3              0.266  0.0596  ...   -18.515     1       0.0363        0.9050   
4              0.618  0.4430  ...    -9.681     1       0.0526        0.4690   
...              ...     ...  ...       ...   ...          ...           ...   
113995         0.172  0.2350  ...   -16.393     1       0.0422        0.6400   
113996         0.174  0.1170  ...   -18.318     0       0.0401        0.9940   
113997         0.629  0.3290  ...   -10.895     0       0.0420        0.8670   
113998         0.587  0.5060  ...   -10.889     1       0.0297        0.3810   
113999         0.526  0.4870  ...   -10.204     0       0.0725        0.6810   

        instrumentalness  liveness  valence    tempo  time_signature  \
0               0.000001    0.3580   0.7150   87.917               4   
1               0.000006    0.1010   0.2670   77.489               4   
2               0.000000    0.1170   0.1200   76.332               4   
3               0.000071    0.1320   0.1430  181.740               3   
4               0.000000    0.0829   0.1670  119.949               4   
...                  ...       ...      ...      ...             ...   
113995          0.928000    0.0863   0.0339  125.995               5   
113996          0.97

In [7]:
df.describe(include=['O'])

track_id       artists                  album_name  \
count                    89740         89740                       89740   
unique                   89740         31437                       46589   
top     5SuOikwiRyPMVoIQDJUgSV  George Jones  The Complete Hank Williams   
freq                         1           260                         110   

                               track_name track_genre  
count                               89740       89740  
unique                              73608         113  
top     Rockin' Around The Christmas Tree    acoustic  
freq                                   48        1000

Based on the description, George Jones is the most frequent artist that appears in the data, and acoustic is the genre with highest frequency. 

Now, we are finding out the most popular track, the genre with tracks that have highest popularity scores, as well as the artists.

In [8]:
most_popular_track = df['popularity'].idxmax()
print(df.loc[most_popular_track])
least_popular_track = df['popularity'].idxmin()
df.loc[least_popular_track]

Unnamed: 0                              20001
track_id               3nqQXoyQOWXiESFLlDF1hG
artists                  Sam Smith;Kim Petras
album_name          Unholy (feat. Kim Petras)
track_name          Unholy (feat. Kim Petras)
popularity                                100
duration_ms                            156943
explicit                                False
danceability                            0.714
energy                                  0.472
key                                         2
loudness                               -7.375
mode                                        1
speechiness                            0.0864
acousticness                            0.013
instrumentalness                     0.000005
liveness                                0.266
valence                                 0.238
tempo                                 131.121
time_signature                              4
track_genre                             dance
Name: 20001, dtype: object


Unnamed: 0                              23
track_id            0BUuuEvNa5T4lMaewyiudB
artists                         Jason Mraz
album_name                   Coffee Moment
track_name                93 Million Miles
popularity                               0
duration_ms                         216386
explicit                             False
danceability                         0.572
energy                               0.454
key                                      3
loudness                           -10.286
mode                                     1
speechiness                         0.0258
acousticness                         0.477
instrumentalness                  0.000014
liveness                            0.0974
valence                              0.515
tempo                              140.182
time_signature                           4
track_genre                       acoustic
Name: 23, dtype: object

Currently, the most popular song is Unholy (feat. Kim Petras) by Sam Smith and Kim Petras, with a popularity score of 100. On the other hand, the least popular song in this dataset is 93 Million Miles by Jason Mraz with 0 popularity.

Now we see how many genres are there in this dataset, and which one is the most/least popular one, based on the genres' mean popularity scores.

In [9]:
df.groupby('track_genre')['popularity'].mean().sort_values(ascending=False)

track_genre
k-pop             59.423581
pop-film          59.096933
metal             56.422414
chill             53.738683
latino            51.788945
                    ...    
detroit-techno    11.130753
latin              9.855072
jazz               9.790076
romance            3.549779
iranian            2.224696
Name: popularity, Length: 113, dtype: float64

There are a total of 113 genres in this dataset. The mean popularity for each one is calculated and sorted in a descending order, which highlights the following result: 
* K-pop is the most popular genre, with an average popularity score of 59.423581
* Iranian is the least popular genre, with an average popularity score of 2.224696

We examine the correlation between these variables
* Popularity and Instrumentalness
* Popularity and Danceability
* Popularity and Liveness
* Popularity and Energy
* Popularity and Duration
* Popularity and Acousticness
* Popularity and Tempo
...

In [10]:
# Popularity and Audio features
selected_columns = ['instrumentalness', 'danceability', 'liveness', 'energy', 'acousticness', 'valence', 'tempo', 'duration_ms']
correlation_table = pd.DataFrame(columns=['Audio Feature', 'Correlation'])
for feature in selected_columns:
    correlation = df['popularity'].corr(df[feature])
    correlation_table.loc[len(correlation_table)] = [feature, correlation]
correlation_table

Audio Feature  Correlation
0  instrumentalness    -0.127477
1      danceability     0.064275
2          liveness    -0.013857
3            energy     0.013725
4      acousticness    -0.038828
5           valence    -0.011508
6             tempo     0.007280
7       duration_ms    -0.023157

We want to address the collinearity issues in our dataset. First, we create a correlation matrix between the predictors.

In [11]:
# Convert the explicit predictor to 0 and 1 (true and false)
df['explicit'] = df['explicit'].astype(int)
df

/tmp/ipykernel_21/2043718031.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['explicit'] = df['explicit'].astype(int)


Unnamed: 0                track_id                 artists  \
0                0  5SuOikwiRyPMVoIQDJUgSV             Gen Hoshino   
1                1  4qPNDBW1i3p13qLCt0Ki3A            Ben Woodward   
2                2  1iJBSr7s7jYXzM8EGcbK5b  Ingrid Michaelson;ZAYN   
3                3  6lfxq3CG4xtTiEg7opyCyx            Kina Grannis   
4                4  5vjLSffimiIP26QG5WcN2K        Chord Overstreet   
...            ...                     ...                     ...   
113995      113995  2C3TZjDRiAzdyViavDJ217           Rainy Lullaby   
113996      113996  1hIz5L4IB9hN3WRYPOCGPw           Rainy Lullaby   
113997      113997  6x8ZfSoqDjuNa5SVP5QjvX           Cesária Evora   
113998      113998  2e6sXL2bYv4bSz6VTdnfLs        Michael W. Smith   
113999      113999  2hETkH7cOfqmz3LqZDHZf5           Cesária Evora   

                                               album_name  \
0                                                  Comedy   
1                                        Ghost (Acoustic)   
2                                          To Begin Again   
3       Crazy Rich Asians (Original Motion Picture Sou...   
4                                                 Hold On   
...                                                   ...   
113995  #mindfulness - Soft Rain for Mindful Meditatio...   
113996  #mindfulness - Soft Rain for Mindful Meditatio...   
113997                                            Best Of   
113998                                  Change Your World   
113999                                     Miss Perfumado   

                        track_name  popularity  duration_ms  explicit  \
0                           Comedy          73       230666         0   
1                 Ghost - Acoustic          55       149610         0   
2                   To Begin Again          57       210826         0   
3       Can't Help Falling In Love          71       201933         0   
4                          Hold On          82       198853         0   
...                            ...         ...          ...       ...   
113995         Sleep My Little Boy          21       384999         0   
113996            Water Into Light          22       385000         0   
113997              Miss Perfumado          22       271466         0   
113998                     Friends          41       283893         0   
113999                   Barbincor          22       241826         0   

        danceability  energy  ...  loudness  mode  speechiness  acousticness  \
0              0.676  0.4610  ...    -6.746     0       0.1430        0.0322   
1              0.420  0.1660  ...   -17.235     1       0.0763        0.9240   
2              0.438  0.3590  ...    -9.734     1       0.0557        0.2100   
3              0.266  0.0596  ...   -18.515     1       0.0363        0.9050   
4              0.618  0.4430  ...    -9.681     1       0.0526        0.4690   
...              ...     ...  ...       ...   ...          ...           ...   
113995         0.172  0.2350  ...   -16.393     1       0.0422        0.6400   
113996         0.174  0.1170  ...   -18.318     0       0.0401        0.9940   
113997         0.629  0.3290  ...   -10.895     0       0.0420        0.8670   
113998         0.587  0.5060  ...   -10.889     1       0.0297        0.3810   
113999         0.526  0.4870  ...   -10.204     0       0.0725        0.6810   

        instrumentalness  liveness  valence    tempo  time_signature  \
0               0.000001    0.3580   0.7150   87.917               4   
1               0.000006    0.1010   0.2670   77.489               4   
2               0.000000    0.1170   0.1200   76.332               4   
3               0.000071    0.1320   0.1430  181.740               3   
4               0.000000    0.0829   0.1670  119.949               4   
...                  ...       ...      ...      ...             ...   
113995          0.928000    0.0863   0.0339  125.995               5   
113996          0.97

In [12]:
# Select quantitative predictors:
quantitative_predictors = df.select_dtypes(include=['int64', 'float64'])
correlation_matrix = quantitative_predictors.corr()
correlation_matrix

Unnamed: 0  popularity  duration_ms  explicit  danceability  \
Unnamed: 0          1.000000    0.036328    -0.029044 -0.065412     -0.025993   
popularity          0.036328    1.000000    -0.023157  0.054898      0.064275   
duration_ms        -0.029044   -0.023157     1.000000 -0.065373     -0.064139   
explicit           -0.065412    0.054898    -0.065373  1.000000      0.119115   
danceability       -0.025993    0.064275    -0.064139  0.119115      1.000000   
energy             -0.061693    0.013725     0.063258  0.100696      0.143913   
key                -0.009098    0.003441     0.011297  0.006634      0.035116   
loudness           -0.043376    0.071674     0.000360  0.109048      0.271838   
mode               -0.006199   -0.016238    -0.039378 -0.033853     -0.062639   
speechiness        -0.086852   -0.047086    -0.060751  0.328201      0.109519   
acousticness        0.086952   -0.038828    -0.108478 -0.091508     -0.177040   
instrumentalness   -0.048577   -0.127477     0.117372 -0.112077     -0.194750   
liveness            0.048416   -0.013857     0.008109  0.041329     -0.132237   
valence             0.034995   -0.011508    -0.147018 -0.002709      0.492578   
tempo              -0.032262    0.007280     0.026004 -0.004070     -0.021329   
time_signature     -0.030159    0.036898     0.022608  0.037128      0.210043   

                    energy       key  loudness      mode  speechiness  \
Unnamed: 0       -0.061693 -0.009098 -0.043376 -0.006199    -0.086852   
popularity        0.013725  0.003441  0.071674 -0.016238    -0.047086   
duration_ms       0.063258  0.011297  0.000360 -0.039378    -0.060751   
explicit          0.100696  0.006634  0.109048 -0.033853     0.328201   
danceability      0.143913  0.035116  0.271838 -0.062639     0.109519   
energy            1.000000  0.046335  0.758774 -0.075018     0.141332   
key               0.046335  1.000000  0.035903 -0.142090     0.017761   
loudness          0.758774  0.035903  1.000000 -0.037111     0.055295   
mode             -0.075018 -0.142090 -0.037111  1.000000    -0.040802   
speechiness       0.141332  0.017761  0.055295 -0.040802     1.000000   
acousticness     -0.732569 -0.045165 -0.582664  0.091038     0.005124   
instrumentalness -0.181038 -0.005281 -0.434719 -0.052588    -0.099659   
liveness          0.188940 -0.002580  0.077238  0.016766     0.225100   
valence           0.256317  0.026253  0.289095  0.025583     0.032935   
tempo             0.258629  0.008272  0.228660 -0.001895     0.004033   
time_signature    0.186534  0.014800  0.193695 -0.023900    -0.006198   

                  acousticness  instrumentalness  liveness   valence  \
Unnamed: 0            0.086952         -0.048577  0.048416  0.034995   
popularity           -0.038828         -0.127477 -0.013857 -0.011508   
duration_ms          -0.108478          0.117372  0.008109 -0.147018   
explicit             -0.091508         -0.112077  0.041329 -0.002709   
danceability         -0.177040         -0.194750 -0.132237  0.492578   
energy               -0.732569         -0.181038  0.188940  0.256317   
key                  -0.045165         -0.005281 -0.002580  0.026253   
loudness             -0.582664         -0.434719  0.077238  0.289095   
mode                  0.091038         -0.052588  0.016766  0.025583   
speechiness           0.005124         -0.099659  0.225100  0.032935   
acousticness          1.000000          0.096072 -0.019073 -0.103781   
instrumentalness      0.096072          1.000000 -0.087157 -0.330756   
liveness             -0.019073         -0.087157  1.000000  0.012322   
valence              -0.103781         -0.330756  0.012322  1.000000   
tempo                -0.220252         -0.056885 -0.006113  0.090617   
time_signature       -0.174057         -0.082235 -0.030785  0.135523   

                     tempo  time_signature  
Unnamed: 0       -0.032262       -0.030159  
popularity        0.007280        0.036898  
duration_ms       0.026004        0.02

From this matrix, we observe a high positive correlation between danceability and loudness (0.7588), which means that the louder the audio, the higher the danceability of the music. Moreover, the strong negative correlation between energy and acousticness is noticeable (-0.7326). 
Thus, we will combine the loudness and dancebality to 1 new feature called 'loud_dance', which is the average of both predictors; and removinng acousticness from our dataset, as it can be inferred from the energy feature.

In [13]:
df['loud_dance'] = (df['danceability'] + df['loudness']) / 2
df

/tmp/ipykernel_21/1484307716.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['loud_dance'] = (df['danceability'] + df['loudness']) / 2


Unnamed: 0                track_id                 artists  \
0                0  5SuOikwiRyPMVoIQDJUgSV             Gen Hoshino   
1                1  4qPNDBW1i3p13qLCt0Ki3A            Ben Woodward   
2                2  1iJBSr7s7jYXzM8EGcbK5b  Ingrid Michaelson;ZAYN   
3                3  6lfxq3CG4xtTiEg7opyCyx            Kina Grannis   
4                4  5vjLSffimiIP26QG5WcN2K        Chord Overstreet   
...            ...                     ...                     ...   
113995      113995  2C3TZjDRiAzdyViavDJ217           Rainy Lullaby   
113996      113996  1hIz5L4IB9hN3WRYPOCGPw           Rainy Lullaby   
113997      113997  6x8ZfSoqDjuNa5SVP5QjvX           Cesária Evora   
113998      113998  2e6sXL2bYv4bSz6VTdnfLs        Michael W. Smith   
113999      113999  2hETkH7cOfqmz3LqZDHZf5           Cesária Evora   

                                               album_name  \
0                                                  Comedy   
1                                        Ghost (Acoustic)   
2                                          To Begin Again   
3       Crazy Rich Asians (Original Motion Picture Sou...   
4                                                 Hold On   
...                                                   ...   
113995  #mindfulness - Soft Rain for Mindful Meditatio...   
113996  #mindfulness - Soft Rain for Mindful Meditatio...   
113997                                            Best Of   
113998                                  Change Your World   
113999                                     Miss Perfumado   

                        track_name  popularity  duration_ms  explicit  \
0                           Comedy          73       230666         0   
1                 Ghost - Acoustic          55       149610         0   
2                   To Begin Again          57       210826         0   
3       Can't Help Falling In Love          71       201933         0   
4                          Hold On          82       198853         0   
...                            ...         ...          ...       ...   
113995         Sleep My Little Boy          21       384999         0   
113996            Water Into Light          22       385000         0   
113997              Miss Perfumado          22       271466         0   
113998                     Friends          41       283893         0   
113999                   Barbincor          22       241826         0   

        danceability  energy  ...  mode  speechiness  acousticness  \
0              0.676  0.4610  ...     0       0.1430        0.0322   
1              0.420  0.1660  ...     1       0.0763        0.9240   
2              0.438  0.3590  ...     1       0.0557        0.2100   
3              0.266  0.0596  ...     1       0.0363        0.9050   
4              0.618  0.4430  ...     1       0.0526        0.4690   
...              ...     ...  ...   ...          ...           ...   
113995         0.172  0.2350  ...     1       0.0422        0.6400   
113996         0.174  0.1170  ...     0       0.0401        0.9940   
113997         0.629  0.3290  ...     0       0.0420        0.8670   
113998         0.587  0.5060  ...     1       0.0297        0.3810   
113999         0.526  0.4870  ...     0       0.0725        0.6810   

        instrumentalness  liveness  valence    tempo  time_signature  \
0               0.000001    0.3580   0.7150   87.917               4   
1               0.000006    0.1010   0.2670   77.489               4   
2               0.000000    0.1170   0.1200   76.332               4   
3               0.000071    0.1320   0.1430  181.740               3   
4               0.000000    0.0829   0.1670  119.949               4   
...                  ...       ...      ...      ...             ...   
113995          0.928000    0.0863   0.0339  125.995               5   
113996          0.976000    0.1050   0.0350   85.239               4   
113997          0.000000    0.0839   0.7430  132.378               4

Split the data into training and testing datasets

In [14]:
# Drop loudness and danceability due to the new combined predictor loud_dance
df = df.drop('loudness', axis=1)
df = df.drop('danceability', axis=1)
df

Unnamed: 0                track_id                 artists  \
0                0  5SuOikwiRyPMVoIQDJUgSV             Gen Hoshino   
1                1  4qPNDBW1i3p13qLCt0Ki3A            Ben Woodward   
2                2  1iJBSr7s7jYXzM8EGcbK5b  Ingrid Michaelson;ZAYN   
3                3  6lfxq3CG4xtTiEg7opyCyx            Kina Grannis   
4                4  5vjLSffimiIP26QG5WcN2K        Chord Overstreet   
...            ...                     ...                     ...   
113995      113995  2C3TZjDRiAzdyViavDJ217           Rainy Lullaby   
113996      113996  1hIz5L4IB9hN3WRYPOCGPw           Rainy Lullaby   
113997      113997  6x8ZfSoqDjuNa5SVP5QjvX           Cesária Evora   
113998      113998  2e6sXL2bYv4bSz6VTdnfLs        Michael W. Smith   
113999      113999  2hETkH7cOfqmz3LqZDHZf5           Cesária Evora   

                                               album_name  \
0                                                  Comedy   
1                                        Ghost (Acoustic)   
2                                          To Begin Again   
3       Crazy Rich Asians (Original Motion Picture Sou...   
4                                                 Hold On   
...                                                   ...   
113995  #mindfulness - Soft Rain for Mindful Meditatio...   
113996  #mindfulness - Soft Rain for Mindful Meditatio...   
113997                                            Best Of   
113998                                  Change Your World   
113999                                     Miss Perfumado   

                        track_name  popularity  duration_ms  explicit  energy  \
0                           Comedy          73       230666         0  0.4610   
1                 Ghost - Acoustic          55       149610         0  0.1660   
2                   To Begin Again          57       210826         0  0.3590   
3       Can't Help Falling In Love          71       201933         0  0.0596   
4                          Hold On          82       198853         0  0.4430   
...                            ...         ...          ...       ...     ...   
113995         Sleep My Little Boy          21       384999         0  0.2350   
113996            Water Into Light          22       385000         0  0.1170   
113997              Miss Perfumado          22       271466         0  0.3290   
113998                     Friends          41       283893         0  0.5060   
113999                   Barbincor          22       241826         0  0.4870   

        key  mode  speechiness  acousticness  instrumentalness  liveness  \
0         1     0       0.1430        0.0322          0.000001    0.3580   
1         1     1       0.0763        0.9240          0.000006    0.1010   
2         0     1       0.0557        0.2100          0.000000    0.1170   
3         0     1       0.0363        0.9050          0.000071    0.1320   
4         2     1       0.0526        0.4690          0.000000    0.0829   
...     ...   ...          ...           ...               ...       ...   
113995    5     1       0.0422        0.6400          0.928000    0.0863   
113996    0     0       0.0401        0.9940          0.976000    0.1050   
113997    0     0       0.0420        0.8670          0.000000    0.0839   
113998    7     1       0.0297        0.3810          0.000000    0.2700   
113999    1     0       0.0725        0.6810          0.000000    0.0893   

        valence    tempo  time_signature  track_genre  loud_dance  
0        0.7150   87.917               4     acoustic     -3.0350  
1        0.2670   77.489               4     acoustic     -8.4075  
2        0.1200   76.332               4     acoustic     -4.6480  
3        0.1430  181.740               3     acoustic     -9.1245  
4        0.1670  119.949               4     acoustic     -4.5315  
...         ...      ...             ...          ...         ...  
113995   0.0339  125.995               5  world-music     -8.1105  
1139

In [15]:
# Drop acousticness due to its high correlation with energy
df = df.drop('acousticness', axis=1)
df

Unnamed: 0                track_id                 artists  \
0                0  5SuOikwiRyPMVoIQDJUgSV             Gen Hoshino   
1                1  4qPNDBW1i3p13qLCt0Ki3A            Ben Woodward   
2                2  1iJBSr7s7jYXzM8EGcbK5b  Ingrid Michaelson;ZAYN   
3                3  6lfxq3CG4xtTiEg7opyCyx            Kina Grannis   
4                4  5vjLSffimiIP26QG5WcN2K        Chord Overstreet   
...            ...                     ...                     ...   
113995      113995  2C3TZjDRiAzdyViavDJ217           Rainy Lullaby   
113996      113996  1hIz5L4IB9hN3WRYPOCGPw           Rainy Lullaby   
113997      113997  6x8ZfSoqDjuNa5SVP5QjvX           Cesária Evora   
113998      113998  2e6sXL2bYv4bSz6VTdnfLs        Michael W. Smith   
113999      113999  2hETkH7cOfqmz3LqZDHZf5           Cesária Evora   

                                               album_name  \
0                                                  Comedy   
1                                        Ghost (Acoustic)   
2                                          To Begin Again   
3       Crazy Rich Asians (Original Motion Picture Sou...   
4                                                 Hold On   
...                                                   ...   
113995  #mindfulness - Soft Rain for Mindful Meditatio...   
113996  #mindfulness - Soft Rain for Mindful Meditatio...   
113997                                            Best Of   
113998                                  Change Your World   
113999                                     Miss Perfumado   

                        track_name  popularity  duration_ms  explicit  energy  \
0                           Comedy          73       230666         0  0.4610   
1                 Ghost - Acoustic          55       149610         0  0.1660   
2                   To Begin Again          57       210826         0  0.3590   
3       Can't Help Falling In Love          71       201933         0  0.0596   
4                          Hold On          82       198853         0  0.4430   
...                            ...         ...          ...       ...     ...   
113995         Sleep My Little Boy          21       384999         0  0.2350   
113996            Water Into Light          22       385000         0  0.1170   
113997              Miss Perfumado          22       271466         0  0.3290   
113998                     Friends          41       283893         0  0.5060   
113999                   Barbincor          22       241826         0  0.4870   

        key  mode  speechiness  instrumentalness  liveness  valence    tempo  \
0         1     0       0.1430          0.000001    0.3580   0.7150   87.917   
1         1     1       0.0763          0.000006    0.1010   0.2670   77.489   
2         0     1       0.0557          0.000000    0.1170   0.1200   76.332   
3         0     1       0.0363          0.000071    0.1320   0.1430  181.740   
4         2     1       0.0526          0.000000    0.0829   0.1670  119.949   
...     ...   ...          ...               ...       ...      ...      ...   
113995    5     1       0.0422          0.928000    0.0863   0.0339  125.995   
113996    0     0       0.0401          0.976000    0.1050   0.0350   85.239   
113997    0     0       0.0420          0.000000    0.0839   0.7430  132.378   
113998    7     1       0.0297          0.000000    0.2700   0.4130  135.960   
113999    1     0       0.0725          0.000000    0.0893   0.7080   79.198   

        time_signature  track_genre  loud_dance  
0                    4     acoustic     -3.0350  
1                    4     acoustic     -8.4075  
2                    4     acoustic     -4.6480  
3                    3     acoustic     -9.1245  
4                    4     acoustic     -4.5315  
...                ...          ...         ...  
113995               5  world-music     -8.1105  
113996               4  world-music     -9.0720  
113997               4  world-music     -5.1330  


The popularity of a song can be related to the number of artists participated in it. For example, 2 artists collaborating might attract more listeners, as each has their own fans. Therefore, we create a new predictor that represents the number of artists on this track.

In [16]:
# The number of artists featured on a track can be represented by the number of occurences of ";"
df['num_artists'] = df['artists'].str.count(';') + 1
df

Unnamed: 0                track_id                 artists  \
0                0  5SuOikwiRyPMVoIQDJUgSV             Gen Hoshino   
1                1  4qPNDBW1i3p13qLCt0Ki3A            Ben Woodward   
2                2  1iJBSr7s7jYXzM8EGcbK5b  Ingrid Michaelson;ZAYN   
3                3  6lfxq3CG4xtTiEg7opyCyx            Kina Grannis   
4                4  5vjLSffimiIP26QG5WcN2K        Chord Overstreet   
...            ...                     ...                     ...   
113995      113995  2C3TZjDRiAzdyViavDJ217           Rainy Lullaby   
113996      113996  1hIz5L4IB9hN3WRYPOCGPw           Rainy Lullaby   
113997      113997  6x8ZfSoqDjuNa5SVP5QjvX           Cesária Evora   
113998      113998  2e6sXL2bYv4bSz6VTdnfLs        Michael W. Smith   
113999      113999  2hETkH7cOfqmz3LqZDHZf5           Cesária Evora   

                                               album_name  \
0                                                  Comedy   
1                                        Ghost (Acoustic)   
2                                          To Begin Again   
3       Crazy Rich Asians (Original Motion Picture Sou...   
4                                                 Hold On   
...                                                   ...   
113995  #mindfulness - Soft Rain for Mindful Meditatio...   
113996  #mindfulness - Soft Rain for Mindful Meditatio...   
113997                                            Best Of   
113998                                  Change Your World   
113999                                     Miss Perfumado   

                        track_name  popularity  duration_ms  explicit  energy  \
0                           Comedy          73       230666         0  0.4610   
1                 Ghost - Acoustic          55       149610         0  0.1660   
2                   To Begin Again          57       210826         0  0.3590   
3       Can't Help Falling In Love          71       201933         0  0.0596   
4                          Hold On          82       198853         0  0.4430   
...                            ...         ...          ...       ...     ...   
113995         Sleep My Little Boy          21       384999         0  0.2350   
113996            Water Into Light          22       385000         0  0.1170   
113997              Miss Perfumado          22       271466         0  0.3290   
113998                     Friends          41       283893         0  0.5060   
113999                   Barbincor          22       241826         0  0.4870   

        key  mode  speechiness  instrumentalness  liveness  valence    tempo  \
0         1     0       0.1430          0.000001    0.3580   0.7150   87.917   
1         1     1       0.0763          0.000006    0.1010   0.2670   77.489   
2         0     1       0.0557          0.000000    0.1170   0.1200   76.332   
3         0     1       0.0363          0.000071    0.1320   0.1430  181.740   
4         2     1       0.0526          0.000000    0.0829   0.1670  119.949   
...     ...   ...          ...               ...       ...      ...      ...   
113995    5     1       0.0422          0.928000    0.0863   0.0339  125.995   
113996    0     0       0.0401          0.976000    0.1050   0.0350   85.239   
113997    0     0       0.0420          0.000000    0.0839   0.7430  132.378   
113998    7     1       0.0297          0.000000    0.2700   0.4130  135.960   
113999    1     0       0.0725          0.000000    0.0893   0.7080   79.198   

        time_signature  track_genre  loud_dance  num_artists  
0                    4     acoustic     -3.0350            1  
1                    4     acoustic     -8.4075            1  
2                    4     acoustic     -4.6480            2  
3                    3     acoustic     -9.1245            1  
4                    4     acoustic     -4.5315            1  
...                ...          ...         ...          ...  
113995               5  world-music     -8.1105            